File: project_versions/full_backup_20260328_001355/frontend/src/api/unified.ts
/**
 * MyQuant v10.0.0 - 统一数据API
 * 遵守架构原则：前端通过统一接口获取所有数据
 *
 * > 创建日期: 2026-02-06
 * > 核心原则:
 * > - 前端只调用 /api/v1/data/unified
 * > - 后端UnifiedDataManager内部选择最优数据源
 * > - 自动应用数据格式转换规则
 *
 * 架构流程:
 * 前端组件 → unified.ts → /api/v1/data/unified → UnifiedDataManager → 数据源
 *                                                      ↓
 *                                                 按规则选择:
 *                                                 - XtQuant下载+读取
 *                                                 - 在线获取
 *                                                 - 缓存策略
 *                                                 - L0-L5层级
 */

import { http } from './request'

// ==================== 类型定义 ====================

/**
 * 支持的数据类型
 */
export type UnifiedDataType =
  | 'kline'          // K线数据
  | 'realtime_quote' // 实时行情
  | 'financial'      // 财务数据
  | 'tick'           // 分笔数据

/**
 * K线周期类型（与后端保持一致）
 */
export type KlinePeriod =
  | '1min'  // 1分钟
  | '5min'  // 5分钟
  | '15min' // 15分钟
  | '30min' // 30分钟
  | '60min' // 60分钟
  | 'day'   // 日线
  | 'week'  // 周线
  | 'month' // 月线

/**
 * 统一数据请求接口
 */
export interface UnifiedDataRequest {
  data_type: UnifiedDataType
  symbols: string[]
  params?: Record<string, any>
  use_cache?: boolean  // 是否使用缓存（默认true）
}

/**
 * K线数据项（后端返回格式）
 */
export interface KlineDataItem {
  time: string           // ISO 8601格式: "2026-02-05 15:00:00"
  open: number
  high: number
  low: number
  close: number
  volume: number
  amount?: number
  trading_day_index?: number  // 交易日历索引
}

/**
 * 统一数据响应
 */
export interface UnifiedDataResponse<T = any> {
  code: number           // 200=成功
  data?: T
  message: string
  metadata?: {
    source: string        // 数据源名称
    elapsed_ms: number   // 请求耗时
    data_level?: string  // L0-L5级别
    timestamp: string
  }
}

// ==================== K线数据API ====================

/**
 * 获取K线数据（统一接口）
 *
 * ⭐ 推荐使用：通过UnifiedDataManager自动选择最优数据源
 *
 * 后端处理流程:
 * 1. 接收请求
 * 2. UnifiedDataManager根据规则选择数据源:
 *    - 检查本地缓存（L0-L2）
 *    - XtQuant下载+读取（优先）
 *    - 在线获取（备用）
 * 3. 应用数据格式规则（time字段格式、复权类型等）
 * 4. 返回统一格式数据
 *
 * @param symbol 股票代码（如 "600519.SH"）
 * @param period K线周期
 * @param count 数据条数（默认500，最大10000）
 * @param use_cache 是否使用缓存（默认true）
 * @returns K线数据
 *
 * @example
 * // 获取贵州茅台日K线
 * const data = await getUnifiedKline('600519.SH', 'day', 250)
 *
 * @example
 * // 获取5分钟K线
 * const data = await getUnifiedKline('600519.SH', '5min', 500)
 */
export const getUnifiedKline = async (
  symbol: string,
  period: KlinePeriod,
  count: number = 500,
  use_cache: boolean = true
): Promise<UnifiedDataResponse<KlineDataItem[]>> => {
  const request: UnifiedDataRequest = {
    data_type: 'kline',
    symbols: [symbol],
    params: {
      period,
      count,
      dividend_type: 'front'  // 前复权（遵守数据格式规则）
    },
    use_cache: use_cache
  }

  return http.post<UnifiedDataResponse<KlineDataItem[]>>('/api/v1/data/unified', request)
    .then(response => response.data)
}

/**
 * 获取日K线数据（便捷函数）
 */
export const getDailyKline = (
  symbol: string,
  count: number = 250
): Promise<UnifiedDataResponse<KlineDataItem[]>> => {
  return getUnifiedKline(symbol, 'day', count)
}

/**
 * 获取周K线数据（便捷函数）
 */
export const getWeeklyKline = (
  symbol: string,
  count: number = 120
): Promise<UnifiedDataResponse<KlineDataItem[]>> => {
  return getUnifiedKline(symbol, 'week', count)
}

/**
 * 获取月K线数据（便捷函数）
 */
export const getMonthlyKline = (
  symbol: string,
  count: number = 60
): Promise<UnifiedDataResponse<KlineDataItem[]>> => {
  return getUnifiedKline(symbol, 'month', count)
}

/**
 * 获取60分钟K线数据（便捷函数）
 */
export const get60MinKline = (
  symbol: string,
  count: number = 120
): Promise<UnifiedDataResponse<KlineDataItem[]>> => {
  return getUnifiedKline(symbol, '60min', count)
}

/**
 * 获取5分钟K线数据（便捷函数）
 */
export const get5MinKline = (
  symbol: string,
  count: number = 500
): Promise<UnifiedDataResponse<KlineDataItem[]>> => {
  return getUnifiedKline(symbol, '5min', count)
}

/**
 * 获取1分钟K线数据（便捷函数）
 */
export const get1MinKline = (
  symbol: string,
  count: number = 500
): Promise<UnifiedDataResponse<KlineDataItem[]>> => {
  return getUnifiedKline(symbol, '1min', count)
}

// ==================== 实时行情API ====================

/**
 * 获取实时行情（统一接口）
 */
export const getRealtimeQuote = async (
  symbols: string[],
  use_cache: boolean = true
): Promise<UnifiedDataResponse<any[]>> => {
  const request: UnifiedDataRequest = {
    data_type: 'realtime_quote',
    symbols,
    use_cache: use_cache
  }

  return http.post<UnifiedDataResponse<any[]>>('/api/v1/data/unified', request)
    .then(response => response.data)
}

// ==================== 数据转换函数 ====================

/**
 * 将统一API返回的K线数据转换为lightweight-charts格式
 *
 * ⚠️ 数据转换规则:
 * - 后端返回: time = "2026-02-05 15:00:00" (ISO 8601字符串)
 * - lightweight-charts需要: time = number (秒级时间戳)
 *
 * @param apiData API返回的K线数据
 * @param includeVolume 是否包含成交量
 * @returns 转换后的数据数组
 */
export function transformUnifiedKlineData(
  apiData: KlineDataItem[],
  includeVolume: boolean = true
): Array<{
  time: number
  open: number
  high: number
  low: number
  close: number
  volume?: number
}> {
  return apiData.map(item => {
    // A股日线用 BusinessDay 格式提取 CST 日期，再转为 UTC 秒级时间戳
    let time: any
    if (typeof item.time === 'string') {
      const datePart = item.time.slice(0, 10) // "2026-03-13"
      const [year, month, day] = datePart.split('-').map(Number)
      time = { year, month, day }
    } else if (typeof item.time === 'number') {
      // 毫秒或秒 → 提取 CST 日期（UTC+8）
      const ms = item.time > 10000000000 ? item.time : item.time * 1000
      const cst = new Date(ms + 8 * 3600 * 1000) // 转 CST
      time = { year: cst.getUTCFullYear(), month: cst.getUTCMonth() + 1, day: cst.getUTCDate() }
    } else {
      const now = new Date()
      time = { year: now.getFullYear(), month: now.getMonth() + 1, day: now.getDate() }
    }
    const timestamp = typeof time === 'object'
      ? Date.UTC(time.year, time.month - 1, time.day) / 1000
      : time

    const transformed: any = {
      time: timestamp, // lightweight-charts使用秒级时间戳
      open: item.open,
      high: item.high,
      low: item.low,
      close: item.close
    }

    if (includeVolume && item.volume !== undefined) {
      transformed.volume = item.volume
    }

    return transformed
  })
}

// ==================== 常量配置 ====================

/**
 * 默认K线配置（遵守K线加载规则）
 */
export const DEFAULT_UNIFIED_KLINE_CONFIG = {
  // 默认数据条数（遵守K线加载规则）
  defaultCount: {
    'month': 60,   // 月K：5年
    'week': 120,   // 周K：2.3年
    'day': 250,    // 日K：1年
    '60min': 120,  // 60分钟K：5天
    '30min': 120,  // 30分钟K：2.5天
    '15min': 120,  // 15分钟K：1.25天
    '5min': 500,   // 5分钟K：1.7天
    '1min': 500    // 1分钟K：0.83天
  },

  // 最大数据条数
  maxCount: 10000,

  // 推荐周期
  recommendedPeriods: [
    { value: 'month', name: '月K', description: '超长期趋势' },
    { value: 'week', name: '周K', description: '中长期趋势' },
    { value: 'day', name: '日K', description: '长期趋势' },
    { value: '60min', name: '60分钟', description: '日内趋势' },
    { value: '5min', name: '5分钟', description: '短线交易' },
    { value: '1min', name: '1分钟', description: '超短线' }
  ]
} as const

File: project_versions/full_backup_20260331_045104/backend/src/myquant/api/dataget/unified.py
"""
统一数据 API 路由

前端通过 /api/v1/data/unified 统一获取所有数据。
后端内部路由到 KlineService（HotDB → LocalDB → 在线源）。
"""

from fastapi import APIRouter, HTTPException
from fastapi.concurrency import run_in_threadpool
from typing import List, Optional
from datetime import datetime
from pydantic import BaseModel, Field
from loguru import logger
import time

from myquant.core.market.services.seamless_service import get_seamless_kline_service
from myquant.core.market.services import get_realtime_market_service
from myquant.core.market.adapters import get_adapter

router = APIRouter(prefix="/data", tags=["统一数据"])


# ─── 请求/响应模型 ────────────────────────────────────────────────────────────

class UnifiedDataRequest(BaseModel):
    data_type: str = Field(..., description="数据类型: kline/realtime_quote/financial/tick")
    symbols: List[str] = Field(..., description="股票代码列表")
    params: Optional[dict] = Field(default_factory=dict, description="请求参数")
    use_cache: bool = Field(True, description="是否使用缓存")


class UnifiedKlineItem(BaseModel):
    """统一K线数据项（与前端 KlineDataItem 对齐）"""
    time: str           # "2026-02-05 15:00:00" 格式
    open: float
    high: float
    low: float
    close: float
    volume: float
    amount: Optional[float] = None
    trading_day_index: Optional[int] = None


class UnifiedDataResponse(BaseModel):
    code: int = 200
    data: Optional[list] = None
    message: str = "success"
    metadata: Optional[dict] = None


# ─── 周期映射 ────────────────────────────────────────────────────────────────

# 前端 unified.ts 周期 → 后端 KlineService 周期
PERIOD_MAP = {
    '1min': '1m',
    '5min': '5m',
    '15min': '15m',
    '30min': '30m',
    '60min': '1h',
    'day': '1d',
    'week': '1w',
    '1w': '1w',  # 周K（RealtimeQuotes.vue 使用）
    'month': '1mon',
    '1M': '1mon',  # 月K（RealtimeQuotes.vue 使用）
}


# ─── 数据转换 ────────────────────────────────────────────────────────────────

def _df_to_unified_items(df) -> List[dict]:
    """将 KlineService 返回的 DataFrame 转换为前端统一格式"""
    import pandas as pd

    if df is None or df.empty:
        return []

    # 去重：按 datetime 去重，保留最后一条
    if 'datetime' in df.columns:
        df = df.drop_duplicates(subset=['datetime'], keep='last')

    items = []
    for _, row in df.iterrows():
        dt_val = row.get('datetime')

        # 处理各种 datetime 类型
        if isinstance(dt_val, (int, float)):
            dt_val = pd.to_datetime(dt_val, unit='ms' if dt_val > 1e11 else 's', errors='coerce')
            if pd.isna(dt_val):
                continue
        elif isinstance(dt_val, str):
            dt_val = pd.to_datetime(dt_val, errors='coerce')
            if pd.isna(dt_val):
                continue

        # 【关键】naive datetime 视为北京时间，返回 UTC 时间戳
        # 北京时间 15:00 = UTC 07:00，相差 8 小时
        if hasattr(dt_val, 'tzinfo') and dt_val.tzinfo is not None:
            # 有时区信息：直接转时间戳
            timestamp_ms = int(dt_val.timestamp() * 1000)
        else:
            # naive datetime：视为北京时间，需要转换为 UTC 时间戳
            # 北京时间 = UTC + 8h，所以 UTC = 北京时间 - 8h
            # 直接用 timestamp() 会在服务器时区下处理，不确定结果
            # 明确转换：视为北京时间，减去 8 小时
            timestamp_s = dt_val.timestamp() - 8 * 3600
            timestamp_ms = int(timestamp_s * 1000)

        vol = float(row.get('volume', 0) or 0)
        if vol == 0:
            continue

        items.append({
            'time': timestamp_ms,  # UTC 时间戳
            'open': float(row.get('open', 0)),
            'high': float(row.get('high', 0)),
            'low': float(row.get('low', 0)),
            'close': float(row.get('close', 0)),
            'volume': vol,
            'amount': float(row.get('amount', 0) or 0),
        })

    return items


# ─── 端点 ─────────────────────────────────────────────────────────────────────

@router.post("/unified", response_model=UnifiedDataResponse)
async def unified_data(request: UnifiedDataRequest):
    """统一数据入口

    前端只调用此接口，后端内部选择最优数据源。
    支持 data_type: kline, realtime_quote, financial, tick
    """
    t0 = time.perf_counter()

    if request.data_type == 'kline':
        return await _handle_kline(request, t0)
    elif request.data_type == 'realtime_quote':
        return await _handle_realtime_quote(request, t0)
    else:
        return UnifiedDataResponse(
            code=400,
            message=f"不支持的数据类型: {request.data_type}",
        )


async def _handle_kline(request: UnifiedDataRequest, t0: float) -> UnifiedDataResponse:
    """处理 K 线数据请求"""
    params = request.params or {}

    # 只取第一个 symbol（当前前端统一接口每次只请求一只股票）
    symbol = request.symbols[0] if request.symbols else None
    if not symbol:
        return UnifiedDataResponse(code=400, message="缺少 symbols 参数")

    # 周期映射
    period = params.get('period', 'day')
    api_period = PERIOD_MAP.get(period, period)

    count = params.get('count', 500)

    logger.info(f"[Unified] kline request: {symbol} period={period}→{api_period} count={count}")

    try:
        service = get_seamless_kline_service()
        df = await run_in_threadpool(
            service.get_kline,
            symbol=symbol,
            period=api_period,
            count=count,
        )

        elapsed_ms = int((time.perf_counter() - t0) * 1000)

        if df is None or df.empty:
            return UnifiedDataResponse(
                code=200,
                data=[],
                message=f"无数据: {symbol} {period}",
                metadata={
                    'source': 'none',
                    'elapsed_ms': elapsed_ms,
                    'timestamp': datetime.now().isoformat(),
                },
            )

        # 检测数据源
        source = 'unknown'
        if 'data_source' in df.columns:
            source = df['data_source'].iloc[0] if not df['data_source'].isna().all() else 'unknown'
        else:
            source = 'kline_service'

        items = _df_to_unified_items(df)

        logger.info(f"[Unified] {symbol} {period}: {len(items)} 条 from {source} in {elapsed_ms}ms")

        return UnifiedDataResponse(
            code=200,
            data=items,
            message="success",
            metadata={
                'source': source,
                'elapsed_ms': elapsed_ms,
                'timestamp': datetime.now().isoformat(),
            },
        )

    except Exception as e:
        elapsed_ms = int((time.perf_counter() - t0) * 1000)
        logger.error(f"[Unified] 获取K线失败: symbol={symbol}, error={e}")
        return UnifiedDataResponse(
            code=500,
            data=[],
            message=str(e),
            metadata={
                'source': 'error',
                'elapsed_ms': elapsed_ms,
                'timestamp': datetime.now().isoformat(),
            },
        )


async def _handle_realtime_quote(request: UnifiedDataRequest, t0: float) -> UnifiedDataResponse:
    """处理实时报价请求"""
    symbols = request.symbols
    if not symbols:
        return UnifiedDataResponse(code=400, message="缺少 symbols 参数")

    logger.info(f"[Unified] realtime_quote request: {symbols}")

    try:
        # 获取实时报价
        service = get_realtime_market_service()
        quotes, data_source = service.get_realtime_quotes(
            codes=symbols,
            use_cache=request.use_cache
        )

        # 补充从 HotDB 计算的财务指标
        for symbol in quotes:
            if data_source != 'tdxquant':
                _fill_indicators_from_hotdb(quotes[symbol], symbol)

        elapsed_ms = int((time.perf_counter() - t0) * 1000)

        return UnifiedDataResponse(
            code=200,
            data=list(quotes.values()),
            message="success",
            metadata={
                'source': data_source,
                'count': len(quotes),
                'elapsed_ms': elapsed_ms,
                'timestamp': datetime.now().isoformat(),
            },
        )

    except Exception as e:
        elapsed_ms = int((time.perf_counter() - t0) * 1000)
        logger.error(f"[Unified] 获取实时报价失败: symbols={symbols}, error={e}")
        return UnifiedDataResponse(
            code=500,
            data=[],
            message=str(e),
            metadata={
                'source': 'error',
                'elapsed_ms': elapsed_ms,
                'timestamp': datetime.now().isoformat(),
            },
        )


def _fill_indicators_from_hotdb(quote: dict, symbol: str) -> None:
    """从 HotDB 补充财务指标（换手率、量比等）

    当 TdxQuant 不可用时，从 HotDB 读取历史数据计算：
    - 换手率 = 当日成交量 / 流通股本
    - 量比 = 当日成交量 / 5日平均成交量
    """
    try:
        import pandas as pd

        adapter = get_adapter('hotdb')
        if adapter is None:
            return

        # 获取最近10天日线数据
        df = adapter.get_kline([symbol], period='1d', count=10)
        if df is None or symbol not in df or df[symbol].empty:
            return

        df_symbol = df[symbol]

        # 当日数据
        today = df_symbol.iloc[-1]
        today_volume = float(today.get('volume', 0))

        # 计算换手率（需要流通股本）
        float_shares = quote.get('float_shares', 0)
        if float_shares > 0 and today_volume > 0:
            turnover_rate = (today_volume / float_shares) * 100
            if quote.get('turnover_rate', 0) == 0:
                quote['turnover_rate'] = round(turnover_rate, 2)

        # 计算量比（当日成交量 / 5日平均成交量）
        if len(df_symbol) >= 6:
            recent_5d_avg = df_symbol.iloc[-6:-1]['volume'].mean()
            if recent_5d_avg > 0 and quote.get('volume_ratio', 0) == 0:
                volume_ratio = today_volume / recent_5d_avg
                quote['volume_ratio'] = round(volume_ratio, 2)

        logger.debug(f"[Unified] {symbol} 从 HotDB 补充指标: "
                    f"换手率={quote.get('turnover_rate', 0)}, "
                    f"量比={quote.get('volume_ratio', 0)}")

    except Exception as e:
        logger.debug(f"[Unified] {symbol} 从 HotDB 补充指标失败: {e}")
